### ISSUE TITLE
Cannot set Linphone Android as default Phone or SMS app

### ISSUE THREAD

**Issue by brianjmurrell** [BelledonneCommunications/linphone-android#1005]
On LineageOS 16.0 (Android Pie), if I open the settings to change default apps, then choose to set the Phone app, I get a list of choices, several of which are SIP clients, but Linphone is not one of them.

Similarly for trying to set the default SMS app.

If I search for a location of some place in Google and then click on it's "Call" item, I'm given a choice of apps, one of which is Linphone, and it opens Linphone with the phone number and allows me to do that Always, which I select and that works in the future.  So that seems to be a work-around to not being able to set the default Phone app to Linphone.

I cannot find any such work-around for SMS though.

On my Huawei Honor 8, running 8.0.0, if I try to change the dialer, Linphone is listed as a choice but selecting it doesn't change it from what it was set to prior.  If I choose a different app, including a SIP app, that change _sticks_.

Is there something more than Linphone should be doing to indicate to Android that it's a Dialer and an SMS app that can be set as defaults?

**Comment by Viish** [comment]
Regarding the call, Linphone shouldn't be displayed on your Huawei like it is on LineageOS (and other Android phones) because it isn't a phone app and doesn't require nor uses the CALL permission. 
Linphone can't be used as a replacement to the SMS app for the same reason.

However it is capable on handling URIs with tel: scheme so that's why you can register it as default app for when a tel: URI is clicked, which happens when you choose a number from a contact for example.

**Comment by Viish** [comment]
Even if a SIP account on the right provider can be used to make GSM calls and convert SIP messages to SMS (which our free service that represents most of our users does not), it might be mind confusing for the user to see Linphone being listed as a potential replacement for the GSM telephony or messaging app.

**Comment by Viish** [comment]
The purpose of being able to replace the default telephony app is to use another application with a different UI but that relies on the GSM network: https://developer.android.com/guide/topics/connectivity/telecom/selfManaged
What you suggested in your first message is the opposite: keep the native look and feel of the telephony app and replace the network stack and protocol.

I don't know how each OS list potential telephony app (I believe it uses the call related permissions in the Manifest or other Telecom API overload) but as we don't need them we don't ask for them.

**Comment by brianjmurrell** [comment]
Not everyone has a mobile plan that includes (affordable, if at all) voice minutes and SMS message allowance.  There are "data only" plans available.  For those people the only choice of a "phone app" and "SMS app" is a VoIP app, which Linphone is perfectly suited to fill.

As I mentioned, there are other SIP apps such as Bria Mobile and Zoiper (to name just a small few), and even some non-SIP VoIP apps such as TextNow that situate themselves to replace the GSM functions on a phone to provide the user with a pure VoIP solution by allowing them to be chosen as default phone and SMS apps.

Seems like this would be just as useful for Linphone to be able to do, no?

**Comment by Viish** [comment]
Linphone could do it, but we have enough bad reviews of the app that says "I dialed a number and it didn't work" that we don't need to add to that the fact that their OS could suggest Linphone as a replacement for their dialing app.
And as I said I don't want our app to ask for permissions we don't need, there are enough of them already.
Finally, this was never asked by any of our clients and/or user except for you.

**Comment by brianjmurrell** [comment]
> Linphone could do it, but we have enough bad reviews of the app that says "I dialed a number and it didn't work"

So instead the app is limited to the lowest-common-denominator of [lack of] intelligence?  There are always going to be stupid and/or ignorant people.  The solution is to educate them, not pander to their ignorance at the expense of the intelligent and knowledgable.

> that we don't need to add to that the fact that their OS could suggest Linphone as a replacement for their dialing app.

I would think of that kind of feature as being optional and advanced, with sufficient warnings.

> And as I said I don't want our app to ask for permissions we don't need, there are enough of them already.

So is that all that is necessary to have Linphone provide itself as an alternative default for phone and sms?  To just ask for some permissions?  I'm sure I could go dig those up, but if you know off hand, any hints on which those are?  I do already build my own Linphone `.apk`, so adding some permission requests is an easy enough local patch to carry.

> Finally, this was never asked by any of our clients and/or user except for you.

So that just makes me the forward thinker who is striving for a better experience.  :-)

Too many people in this world just don't ask enough questions and settle too easily for the status quo instead of striving to make things better.  Then there are those of us that do ask questions and do try to make things better, for everyone.

**Comment by brianjmurrell** [comment]
@Viish Any chance I can ask you to take a peek at my attempt to make Linphone selectable as the default SMS app?  I've done a bunch of Googling and reading and AFAICT, I've added everything I need to, but it's still not showing up as an app I can choose as a default for Messaging.

https://github.com/brianjmurrell/linphone-android/commit/bfe2353dd75a1ba94b013bd9b266054c3a26b064

Anything obvious that you see that I've done wrong?

Much thanks for your time!

**Comment by Viish** [comment]
I'm no expert in how to replace the SMS app but what you did seems to follow the few how-tos I could find quickly on the web:
- https://android-developers.googleblog.com/2013/10/getting-your-sms-apps-ready-for-kitkat.html
- https://www.androidauthority.com/how-to-create-an-sms-app-part-2-724264/

So maybe you have to ask for the permission at runtime and have the user grant them before it can work.
And anyway I'd check when the app starts whether or not it's the default SMS app and if not start the intent to ask to user to change it (the following code is copied/paste from the first link above):
```
        final String myPackageName = getPackageName();
        if (!Telephony.Sms.getDefaultSmsPackage(this).equals(myPackageName)) {
            // App is not default.
            // Show the "not currently set as the default SMS app" interface
            View viewGroup = findViewById(R.id.not_default_app);
            viewGroup.setVisibility(View.VISIBLE);

            // Set up a button that allows the user to change the default SMS app
            Button button = (Button) findViewById(R.id.change_default_app);
            button.setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    Intent intent =
                            new Intent(Telephony.Sms.Intents.ACTION_CHANGE_DEFAULT);
                    intent.putExtra(Telephony.Sms.Intents.EXTRA_PACKAGE_NAME, 
                            myPackageName);
                    startActivity(intent);
                }
            });
        } else {
            // App is the default.
            // Hide the "not currently set as the default SMS app" interface
            View viewGroup = findViewById(R.id.not_default_app);
            viewGroup.setVisibility(View.GONE);
        }
```

**Comment by brianjmurrell** [comment]
Ah ha!  Got it.  One or both of the changes in https://github.com/BelledonneCommunications/linphone-android/commit/511005994ce77760464229785b3befdcc427ffeb enabled it on the default app chooser list.

Interesting message when choosing it though:

_Note: After a reboot, this app can't start until you unlock your phone._

Must be some additional kind of permission needed.  I'll dig into it.

I also noticed Linphone sets: `android:allowBackup="true`"`.  Curious why you don't want it's data being backed up so that setting up a new phone restores the Linphone settings/data.

**Comment by Viish** [comment]
> I also noticed Linphone sets: android:allowBackup="true"`. Curious why you don't want it's data being backed up so that setting up a new phone restores the Linphone settings/data.

If I remember correctly it was because there was an issue when reinstalling the app after uninstall due to files being backuped.
I don't know if that is still necessary.

**Comment by brianjmurrell** [comment]
> _Note: After a reboot, this app can't start until you unlock your phone._

This is indicating a lack of support for Direct Boot in the app and essentially means that Linphone is unable to operate correctly (i.e. receiving calls, messages, posting notifications, etc.) after a reboot on an encrypted device, until the user unlocks the device, providing the key to decrypt the storage.

Direct Boot solves this problem.  [Here](https://www.androidauthority.com/direct-boot-711668-711668/)'s a pretty decent writeup about it.  Should I open a ticket for adding that?

**Comment by Viish** [comment]
> Direct Boot solves this problem. Here's a pretty decent writeup about it. Should I open a ticket for adding that?

We can't be direct boot aware because to received notifications the account(s) must be registered. That would mean storing username & password into non-encrypted storage which isn't safe. Also message received in this mode (and call logs but that is less relevant) wouldn't be stored in local database (unless it was also moved to non-encrypted storage which isn't safe either) and so would be lost upon app restart.
For all of these reasons we aren't ready for that and thus we won't implement this, at least for now, even ever.

Even if your article says your device can reboot at any time without the user being aware, studies shows in average a person looks at it's cellphone every 5 minutes or so, which reduces the importance of the feature.

**Comment by brianjmurrell** [comment]
So, setting as a default SMS app works.  What doesn't though, is that if I go to a contact in a contact app, and click the icon to send an SMS to a contact's number, it opens Linphone, but opens it on the main dialer screen, not in the messaging screen for that contact.

Is that because the `<data android:scheme="smsto" />` is in the `android:name=".dialer.DialerActivity"`?  If so, should it be moved to the `android:name=".chat.ChatActivity"`?  If yes to that, should `<data android:scheme="imto" />` also be moved?

Or am I completely in the weeds on this aspect?

**Comment by Viish** [comment]
I did exactly that yesterday on the feature/kotlin_rewrite branch after I mentioned Linphone could be used to handle click on phone numbers in contacts app and realized we couldn't do it for SMS: https://github.com/BelledonneCommunications/linphone-android/commit/317cc1e1822f752d518ade89fe39786b4b6b8efb

**Comment by brianjmurrell** [comment]
So is all future development being done on the feature/kotlin_rewrite branch exclusively now?  Will master (as it is today) no longer be developed?

**Comment by Viish** [comment]
> So is all future development being done on the feature/kotlin_rewrite branch exclusively now? Will master (as it is today) no longer be developed?

The feature/kotlin_rewrite branch will be merged into master when we will release the SDK 5.0.

> How do you get that to work without being able to change your default SMS app to Linphone

I simply added an intent filter for smsto, sms, mmsto and mms schemes for SEND and SENDTO actions.
If you current SMS app has been set to be the default app, you might have to remove that.

**Comment by brianjmurrell** [comment]
> The feature/kotlin_rewrite branch will be merged into master when we will release the SDK 5.0.

Any idea on the time frame on that?  Just wondering how long bug fixes like this (and the other one to dial a number without interaction when called from another application) will be "in limbo".

I'm assuming feature/kotlin_rewrite is not really yet ready to be used as a daily driver.

**Comment by brianjmurrell** [comment]
> > How do you get that to work without being able to change your default SMS app to Linphone

> I simply added an intent filter for smsto, sms, mmsto and mms schemes for SEND and SENDTO actions.

Which contact app are you using?  Because I have multiple bona-fide SMS/MMS messaging apps installed (the one that came with the ROM, Android Messages) and my Contacts app still doesn't prompt me which app to use when I click a contact to send an SMS.

But is the choice to prompt which app to use even something an app can disable?  I thought it was the Android system that decided which apps to use and when to prompt, etc for intents from other apps.

**Comment by Viish** [comment]
> Any idea on the time frame on that? 

ETA is end of June.

> I'm assuming feature/kotlin_rewrite is not really yet ready to be used as a daily driver.

It hasn't passed all validation tests for a release but I do use it daily and it works quite well so far.

> Which contact app are you using?

I'm using Google contacts app, as well as Messages (from Google) for the SMS.
Android prompts when multiple app are available until one is chosen and user click on "Always use". From that point forward it won't ask again, that's why you have to remove the app as default in it's settings.

**Comment by Viish** [comment]
I also noticed that when I removed the defaults from the Messages app I was prompted with which app to use. However once a updated Linphone from Android Studio, I had to remove again the defaults (maybe Android considers Linphone to have been "removed" and thus grants the default status back to Messages, I don't know).

**Comment by brianjmurrell** [comment]
Currently Google Contacts opens my Linphone for SMS without prompting.  So I opened the app info for my Linphone and cleared defaults.  Did the same for other SMS apps.  Went back to Google Contacts and clicked an smsto icon and it opened my Linphone again, without prompting.

**Comment by Viish** [comment]
As you are using Lineage OS it may behave differently.
Also that might come from your changes to be able to use Linphone as default SMS app.

**Comment by brianjmurrell** [comment]
So yes, it might be a LineageOS thing, to be sure.

But I don't think it's related to my Linphone being an SMS app.  Ultimately I have (at least) 3 SMS apps: my Linphone, the one that came with the ROM, and Android Messages.

For all 3 I cleared their "app open defaults".  That *should* make Android prompt me.  But it doesn't.  I think that's because Android (or LineageOS) don't really want the phone and sms apps to participate in this "choose every time/use always" choice dialog and that's why there is a UI in Settings to choose the default phone and sms apps.

This is how my Huawei Honor 8 behaves also.  It's an 8.0.0 ROM and my LineageOS is a 9.0.x ROM.

What version of Android do you have where you are getting prompted for an SMS app when you click an smsto icon?

**Comment by Viish** [comment]
I have a Samsung running Android 10

**Comment by brianjmurrell** [comment]
There is a LineageOS 17 (Android 10) for my Nexus 5.  Maybe I will try updating to it to see if this is behaviour that changed in 10.  It's also entirely possible that the behaviour I am seeing is the Android default/stock behaviour and Samsung decided that the choice popup was more appropriate and so deviated from stock in that regard.

**Comment by Viish** [comment]
Completely.
Sadly with the "work from home" thing I have a lot less choice of devices to test on.

**Comment by hasanr68** [comment]
@brianjmurrell - this looks interesting. Did you do anything similar for voice calls also?

**Comment by brianjmurrell** [comment]
@hasanr68 Yes.  Even the Google Assistant uses Linphone to make a call when I ask her to call somebody.

**Comment by hasanr68** [comment]
@brianjmurrell thanks. She sounds smart :) How were you able to manage that? Also, is it possible to route calls via the native dialler? I believe ConnectionService or IncallService is used. Thanks.

**Comment by brianjmurrell** [comment]
> How were you able to manage that?

@hasanr68 It's part of the PR I believe.

> Also, is it possible to route calls via the native dialler?

Not with my branch.

> I believe ConnectionService or IncallService is used.

Android P I believe allows for integration of a third-party telephony stack into the native dialer, etc.  I don't have any Android P devices, nor any real itch to implement it.  There doesn't seem to be any desire upstream for Linphone to be anything more than a stand-alone app, that cannot replace the GSM phone or SMS apps.

**Comment by hasanr68** [comment]
@brianjmurrell thanks a lot. I think it would be great if it could replace the GSM phone.

**Comment by brianjmurrell** [comment]
A while ago I [commented](https://github.com/BelledonneCommunications/linphone-android/issues/1005#issuecomment-628094257):

> What doesn't though, is that if I go to a contact in a contact app, and click the icon to send an SMS to a contact's number, it opens Linphone and even opens it in the chat list screen, but not in a chat for the contact. I have to then choose the chat from the list, or start a new one if one doesn't exist.
> 
> Any thoughts on how to get that intent to open into an existing chat for that contact, or create a new one if one doesn't exist?

and @Viish [replied](https://github.com/BelledonneCommunications/linphone-android/issues/1005#issuecomment-628434239):

> I did exactly that yesterday on the feature/kotlin_rewrite branch after I mentioned Linphone could be used to handle click on phone numbers in contacts app and realized we couldn't do it for SMS: 317cc1e

Is that to say/confirm that clicking the "send a message" icon from a contact would not be able to (create if required) the chat for the individual in the contact and that best that can be done is to open linphone in the chat summary screen?  Is this still true on release/4.6.0?

**Comment by Viish** [comment]
> Is that to say/confirm that clicking the "send a message" icon from a contact would not be able to (create if required) the chat for the individual in the contact and that best that can be done is to open linphone in the chat summary screen? Is this still true on release/4.6.0?

Yes and no. 

Currently with 4.6.0 version if Linphone were able to receive a SMS or MMS intent dispatched from Android, it will go to an existing (basic) chat room if one already exists. 
However it doesn't, it will stop at the chat rooms list.

That doesn't mean we won't be able to create a chat room if it doesn't and go inside it in the future, it is just not done yet.

**Comment by anaxonda** [comment]
Hello,

Should it be currently possible to set Linphone as default SMS and Phone app? Running version 5.2.1, android 13, it does not show up in 'Default Apps' settings.

**Comment by LuluZulu123** [comment]
Still nothing as of 2024-04-03 / v.5.2.3 / a14

**Comment by LuluZulu123** [comment]
@Viish actually, I was just about to correct my last comment: after I made one call manually, i was then able to select linphone and "always" when asked with what app to open the directcall widget.
I have not checked to see if it is now set as the default phone handler in the android settings, or if it is just the handler for the tel: uri (don't have the phone with me atm)

**Comment by Viish** [comment]
Yes, that part is implemented, but that doesn't mean it will replace your native dialer, simply that outgoing calls will be made with Linphone (not incoming ones).
